### DISCUSSION TITLE
AuthJS do not redirect to set URL, instead it redirect to Docker container internal IP

### DISCUSSION THREAD

**Discussion by J4v4Scr1pt** [nextauthjs/next-auth#8449] [Help] [answered]
### Environment

  System:
    OS: Windows 10 10.0.19045
    CPU: (12) x64 Intel(R) Core(TM) i7-9850H CPU @ 2.60GHz
    Memory: 7.73 GB / 31.71 GB
  Binaries:
    Node: 20.5.0 - C:\Program Files\nodejs\node.EXE
    npm: 9.8.0 - C:\Program Files\nodejs\npm.CMD
    pnpm: 8.6.12 - ~\AppData\Local\pnpm\pnpm.EXE
  Browsers:
    Edge: Spartan (44.19041.1266.0), Chromium (116.0.1938.54)
    Internet Explorer: 11.0.19041.1566

### Reproduction URL

-

### Describe the issue

When using nextJS and authJS I get redirected to the internal ip-address of Docker instead of set environment variable when trying to access protected urls.

I'm using the app router:
![image](https://github.com/nextauthjs/next-auth/assets/5929055/fe535a2d-390f-4759-8535-cd3454cd002e)

![image](https://github.com/nextauthjs/next-auth/assets/5929055/4a82ddc5-18ed-4531-9fa5-aea11536cf70)
![image](https://github.com/nextauthjs/next-auth/assets/5929055/d64b069c-cd3f-41ee-a0b7-49c3188f40e5)

Tried the suggested solutions in [this ](https://www.reddit.com/r/nextjs/comments/15ryr0u/nextauth_redirect_to_docker_container_ip_instead/)thread but it does not work.

.env:
```
NEXTAUTH_SECRET=secrect
NEXTAUTH_URL=myDomain
```

docker-compose.yaml:
```
version: '3'

services:
  wt-nxt:
    container_name: wt-nxt
    build:
      context: .
      dockerfile: Dockerfile
    restart: always
    ports:
      - 3000:3000
```

Docker:
```
FROM node:18-alpine AS base

FROM base AS builder

WORKDIR /app

COPY package.json yarn.lock* package-lock.json* pnpm-lock.yaml* ./
RUN \
  if [ -f yarn.lock ]; then yarn --frozen-lockfile; \
  elif [ -f package-lock.json ]; then npm ci; \
  elif [ -f pnpm-lock.yaml ]; then yarn global add pnpm && pnpm i; \
  # Allow install without lockfile, so example works even without Node.js installed locally
  else echo "Warning: Lockfile not found. It is recommended to commit lockfiles to version control." && yarn install; \
  fi

COPY . .

RUN \
  if [ -f yarn.lock ]; then yarn build; \
  elif [ -f package-lock.json ]; then npm run build; \
  elif [ -f pnpm-lock.yaml ]; then pnpm build; \
  else yarn build; \
  fi

FROM base AS runner

WORKDIR /app

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs
USER nextjs

COPY --from=builder /app/public ./public

COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static

CMD ["node", "server.js"]
```

auth-api-route:
```
import NextAuth from 'next-auth';
import type { AuthOptions } from 'next-auth';
import CredentialsProvider from 'next-auth/providers/credentials';

const authOptions: AuthOptions = {
	providers: [
		CredentialsProvider({
			credentials: {
				username: { label: 'Username', type: 'text' },
				password: { label: 'Password', type: 'password' },
			},
			async authorize(credentials, req) {

				const res = await fetch('/my/endpoint', {
					method: 'POST',
					body: JSON.stringify(credentials),
				 	headers: { 'Content-Type': 'application/json' },
				});
				const res = await res.json();

			
				if (res.ok && user) {
					return user;
				}
				return null;
			},
		}),
	],
	session: { strategy: 'jwt' },
	pages: {
		signIn: '/Login',
	},
};

const handler = NextAuth(authOptions);

export { handler as GET, handler as POST };

```

middleware:
```
import { withAuth } from 'next-auth/middleware';

export default withAuth({
	callbacks: {
		authorized({ req, token }) {
			return !!token;
		},
	},
	pages: { signIn: '/Login' },
});

export const config = {
	matcher: [
		'/((?!api/|_next/|_static/|_vercel|[\\w-]+\\.\\w+).*)',
	],
};

```

### How to reproduce

Setup NextJS with AuthJS with latest versions.
Then setup a basic docker and dockercompose file.

### Expected behavior

Should redirect to set environment variable.

**Comment by lujun2** [comment]
Encounter the same issue...

After some research, seems like it is related to https://github.com/vercel/next.js/issues/37536

`next-auth` use [`req.nextUrl.origin`](https://github.com/nextauthjs/next-auth/blob/main/packages/next-auth/src/next/middleware.ts#L104) to get the host, but it always return `localhost`


Still under investigation...

**Reply by J4v4Scr1pt** [reply]
Ahhh makes sense thx for the info, I tried to downgrade and it looks like it working locally but the build in my Azure-pipeline got some other error instead 😅.

Edit, works with these versions:
"next": "13.4.3",
"next-auth": "4.22.1",

**Comment by lujun2** [comment]
managed to work when set custom signIn page

```typescript
const authUrl = process.env['NEXTAUTH_URL'] ?? 'http://localhost:3000';

export default withAuth({
  pages: {
    signIn: `${authUrl}/api/auth/signin`
  }
});
```

but I don't think this is a solution ...

**Comment by soulchild** [comment]
This is definitely an issue. All I did was `npm update --save`. All worked well locally, but when I deployed it to our test environment (running with Docker) the login was broken, redirecting my browser to the local docker IP address.

**Reply by J4v4Scr1pt** [reply]
I downgrade the versions to 
"next": "13.4.3",
"next-auth": "4.22.1",

And now everything works.

**Comment by Kiborgik** [comment]
have in .env `NEXTAUTH_URL_INTERNAL=http://localhost:3000`
try add this to docker: 
`ENV HOSTNAME "0.0.0.0"
CMD ["node", "server.js"]`

**Reply by J4v4Scr1pt** [reply]
Did not work, not for me at least 😅

**Reply by VictorGlindasPaf** [reply]
Didn't work for us either

**Reply by axinzi** [reply]
i tryed, and it worked!!! thanks

**Reply by aronei44** [reply]
Its worked. thanks
you must change your NEXTAUTH_URL to http://0.0.0.0:port (container port)
UPDATE
in stag and prod server its not working....
just work in development (local docker)

**Comment by Kiborgik** [comment]
> Did not work, not for me at least 😅
do you have in .env NEXTAUTH_URL_INTERNAL=http://localhost:3000/ thought? (also last versions for both packages)

**Comment by skull8888888** [comment]
any updates on this?

**Reply by J4v4Scr1pt** [reply]
Don't think so, it seems to be about settings.
I have latest version of all packages and it works like a charm for me.
The person that handle our platform removed some trailing slash in k8s, and now it works in prod.

this is my docker file:
```
FROM node:18-alpine AS base

# Step 1. Rebuild the source code only when needed
FROM base AS builder

WORKDIR /app

# Install dependencies based on the preferred package manager
COPY package.json yarn.lock* package-lock.json* pnpm-lock.yaml* ./
# Copy Authentication
COPY .npmrc .
# Omit --production flag for TypeScript devDependencies
RUN \
  if [ -f yarn.lock ]; then yarn --frozen-lockfile; \
  elif [ -f package-lock.json ]; then npm ci; \
  elif [ -f pnpm-lock.yaml ]; then yarn global add pnpm && pnpm i; \
  # Allow install without lockfile, so example works even without Node.js installed locally
  else echo "Warning: Lockfile not found. It is recommended to commit lockfiles to version control." && yarn install; \
  fi
COPY . .
  # if this is the sandbox deployment environment

# Environment variables must be present at build time
# https://github.com/vercel/next.js/discussions/14030

ARG NODE_ENV
ENV NODE_ENV=production
ARG NEXTAUTH_SECRET
ENV NEXTAUTH_SECRET=${NEXTAUTH_SECRET}
ARG NEXTAUTH_URL
ENV NEXTAUTH_URL=${NEXTAUTH_URL}
ARG VERSION
ENV VERSION=${VERSION}
ARG SOURCEVERSION
ENV SOURCEVERSION=${SOURCEVERSION}

# Next.js collects completely anonymous telemetry data about general usage. Learn more here: https://nextjs.org/telemetry
# Uncomment the following line to disable telemetry at build time
ENV NEXT_TELEMETRY_DISABLED 1

# Build Next.js based on the preferred package manager
RUN \
  if [ -f yarn.lock ]; then yarn build; \
  elif [ -f package-lock.json ]; then npm run build; \
  elif [ -f pnpm-lock.yaml ]; then pnpm build; \
  else yarn build; \
  fi

# Note: It is not necessary to add an intermediate step that does a full copy of `node_modules` here

# Step 2. Production image, copy all the files and run next
FROM base AS runner

WORKDIR /app

# Don't run production as root
RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs

COPY --from=builder /app/public ./public

# Set the correct permission for prerender cache
RUN mkdir .next
RUN chown nextjs:nodejs .next

# Automatically leverage output traces to reduce image size
# https://nextjs.org/docs/advanced-features/output-file-tracing
COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static

# Environment variables must be redefined at run time
ARG NODE_ENV
ENV NODE_ENV=production
ARG NEXTAUTH_SECRET
ENV NEXTAUTH_SECRET=${NEXTAUTH_SECRET}
ARG NEXTAUTH_URL
ENV NEXTAUTH_URL=${NEXTAUTH_URL}
ARG VERSION
ENV VERSION=${VERSION}
ARG SOURCEVERSION
ENV SOURCEVERSION=${SOURCEVERSION}

# Uncomment the following line to disable telemetry at run time
ENV NEXT_TELEMETRY_DISABLED 1

USER nextjs

CMD ["node", "server.js"]
```

**Comment by akutruff** [comment]
I just hit this too.  Anyone have any idea?

**Comment by abdulqadeerqureshi** [comment]
Downgrade nextjs back to v13.4.12 and everything works fine. Plus I also added `ENV HOSTNAME 0.0.0.0` to my Dockerfile.

**Comment by aronei44** [comment]
try this. work for me (production too)

i have some problem. the first one is next-auth always redirect to internal docker ip. so, i do this
1. add env in dockerfile
```
ENV HOSTNAME 0.0.0.0
```

now, the redirect is going to ip given. not worked with DNS. so i do this

2. set NEXTAUTH_URL to your domain

example: 
```
NEXTAUTH_URL='https://yourdomain.com'
```
its worked but 'bad gateway' appear. after some search, i found this problem is from NGINX. so i do this

3. setup NGINX

add this into your nginx.conf in http section

```
http {
# another setup

proxy_buffers 8 16k;
proxy_buffer_size 32k;
}
```

i hope this work for you too

**Comment by soulchild** [comment]
Shouldn't this be a bug instead of a discussion, @balazsorban44? Especially, given that it used to work without applying a band-aid? I've worked with a lot of different frameworks over the years, but never have I had to explicitly set my hostname to `0.0.0.0` to make an app work… Just feels strange.

**Comment by xfathurrahman** [comment]
any update for this issuess?

**Comment by masterjanic** [comment]
Same issue here using `"next-auth": "^5.0.0-beta.15"` and `"next": "^14.1.1"`. I have setup `ENV HOSTNAME` and `NEXTAUTH_URL / AUTH_URL` which sets the OAuth redirect url to the correct path but once signed in with any provider it redirects to the internal Docker address which is `0.0.0.0`.

**Reply by tetsuo-keystone** [reply]
I discovered the source of this issue for me, which was that I had the `AUTH_URL` value set in the `.env` file, and that value wasn't available at all times (I tested this by outputting the url and baseUrl values from the redirect callback in the auth config).  The fix was to set the value explicitly as an environment variable on the container in GCP.

Basic setup:
GCP cloudrun, `"next-auth": "^5.0.0-beta.4"`, `"next": "^14.0.4"`

I suspect the issue is when the values from the `.env` file become available.

**Reply by adamspotlite** [reply]
@tetsuo-keystone I also use cloudrun and am experiencing this issue. Are you using `ENV HOSTNAME` in your `Dockerfile`? Is this to say that we need to explictly set `AUTH_URL` as well during build? Can that be done at runtime instead? Also, is `AUTH_URL` just the domain mapping you applied + `api/auth`?

**Reply by tetsuo-keystone** [reply]
Yes, I have `ENV HOSTNAME "0.0.0.0"` in my Dockerfile.

I don't set the `AUTH_URL` value in the .env files at all, I only have it set on the cloudrun container as an environment variable (not sure this makes a difference, also I haven't tested this setup too extensively yet but this is correct to my knowledge right now)

I don't think the `AUTH_URL` can be set at runtime, based on the behaviour I'm seeing.

My `AUTH_URL` value is set to `{custom domain}/api/auth`, so yes!

**Reply by Nishidh25** [reply]
Hi Can you share your docker file and .env config
I have tried everything but i cannot get this to work. In app engine or cloud run 
I have also tried all the combinations.

Would also help if you can share a working repo

Currently running `"next-auth": "^5.0.0-beta.4", "next": "^14.1.0"` 

.env
```AUTH_SECRET=2 

AZURE_AD_CLIENT_ID=""
AZURE_AD_CLIENT_SECRET=""
AZURE_AD_TENANT_ID=""

GOOGLE_CLIENT_ID=""
GOOGLE_CLIENT_SECRET=""


DIALOGFLOW_PROJECT_ID=""
DIALOGFLOW_KEY_FILE="

GOOGLE_APPLICATION_CREDENTIALS=""
```

Dockerfile
```FROM node:20-alpine AS base

FROM base AS deps
libc6-compat might be needed.
RUN apk add --no-cache libc6-compat
WORKDIR /app

COPY package.json yarn.lock* package-lock.json* pnpm-lock.yaml* ./
RUN \
  if [ -f yarn.lock ]; then yarn --frozen-lockfile; \
  elif [ -f package-lock.json ]; then npm ci; \
  elif [ -f pnpm-lock.yaml ]; then corepack enable pnpm && pnpm i --frozen-lockfile; \
  else echo "Lockfile not found." && exit 1; \
  fi


# Rebuild the source code only when needed
FROM base AS builder
WORKDIR /app
COPY --from=deps /app/node_modules ./node_modules
COPY . .


ENV HOSTNAME "0.0.0.0"
ENV PORT 8080
ENV NEXT_TELEMETRY_DISABLED 1
ENV AUTH_SECRET=
ENV AZURE_AD_CLIENT_ID=""
ENV AZURE_AD_CLIENT_SECRET=""
ENV AZURE_AD_TENANT_ID=""
ENV GOOGLE_CLIENT_ID=""
ENV GOOGLE_CLIENT_SECRET=""
ENV DIALOGFLOW_PROJECT_ID=""
ENV DIALOGFLOW_KEY_FILE=""
ENV GOOGLE_APPLICATION_CREDENTIALS=""
ENV AUTH_URL="https://my-production-url/api/auth"


RUN \
  if [ -f yarn.lock ]; then yarn run build; \
  elif [ -f package-lock.json ]; then npm run build; \
  elif [ -f pnpm-lock.yaml ]; then corepack enable pnpm && pnpm run build; \
  else echo "Lockfile not found." && exit 1; \
  fi

FROM base AS runner
WORKDIR /app
ENV NODE_ENV production
ENV NEXT_TELEMETRY_DISABLED 1
ENV AUTH_SECRET=
ENV AZURE_AD_CLIENT_ID=""
ENV AZURE_AD_CLIENT_SECRET=""
ENV AZURE_AD_TENANT_ID=""
ENV GOOGLE_CLIENT_ID=""
ENV GOOGLE_CLIENT_SECRET=""
ENV DIALOGFLOW_PROJECT_ID=""
ENV DIALOGFLOW_KEY_FILE=""
ENV GOOGLE_APPLICATION_CREDENTIALS=""
ENV AUTH_URL="https://my-production-url/api/auth"
RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs
COPY --from=builder /app/public ./public
# Set the correct permission for prerender cache
RUN mkdir .next
RUN chown nextjs:nodejs .next
# Automatically leverage output traces to reduce image size
# https://nextjs.org/docs/advanced-features/output-file-tracing
# copy the public folder from the project as this is not included in the build process
COPY --from=builder --chown=nextjs:nodejs /app/public ./public
# copy the standalone folder inside the .next folder generated from the build process
COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
# copy the static folder inside the .next folder generated from the build process
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static
USER nextjs
# set hostname to localhost
ENV HOSTNAME "0.0.0.0"
ENV PORT 8080
EXPOSE 8080
# server.js is created by next build from the standalone output
# https://nextjs.org/docs/pages/api-reference/next-config-js/output
CMD ["node", "server.js"]
```


I have been stuck here for almost a month: almost thinking about switching my development to Nodejs and use some other well known auth

**Reply by adamspotlite** [reply]
Hey, I was able to get this going with @tetsuo-keystone 's suggestion.

tldr: add `ENV HOSTNAME "0.0.0.0"` and ensure `AUTH_URL` is set on the container runtime.

The setup that worked:

```docker
FROM node:20-alpine AS base
 
FROM base AS builder
RUN apk add --no-cache libc6-compat
RUN apk update
RUN npm install -g pnpm

# Set working directory
WORKDIR /app
RUN npm install -g turbo
COPY . .
RUN turbo prune @spotlite/nextjs --docker
 
# Add lockfile and package.json's of isolated subworkspace
FROM base AS installer
RUN apk add --no-cache libc6-compat
RUN apk update
RUN npm install -g pnpm

WORKDIR /app
 
# First install the dependencies (as they change less often)
COPY .gitignore .gitignore
COPY --from=builder /app/out/json/ .
COPY --from=builder /app/out/pnpm-lock.yaml ./pnpm-lock.yaml
RUN pnpm install
 
# Build the project
COPY --from=builder /app/out/full/ .
RUN CI=true SKIP_ENV_VALIDATION=true pnpm turbo run build --filter=@spotlite/nextjs...
 
FROM base AS runner
WORKDIR /app
 
# Don't run production as root
RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs
USER nextjs
 
COPY --from=installer /app/apps/nextjs/next.config.js .
COPY --from=installer /app/apps/nextjs/package.json .
 
# Automatically leverage output traces to reduce image size
# https://nextjs.org/docs/advanced-features/output-file-tracing
COPY --from=installer --chown=nextjs:nodejs /app/apps/nextjs/.next/standalone ./
COPY --from=installer --chown=nextjs:nodejs /app/apps/nextjs/.next/static ./apps/nextjs/.next/static
COPY --from=installer --chown=nextjs:nodejs /app/apps/nextjs/public ./apps/nextjs/public
 
EXPOSE 3000
ENV PORT 3000
ENV HOSTNAME "0.0.0.0"

CMD node apps/nextjs/server.js
```

**NOTE**: This is from the [t3-turbo repo](https://github.com/t3-oss/create-t3-turbo) - I can't speak to how things are different in a setting outside of that.

Additionally, I deploy to `Cloud Run` and pass `AUTH_URL` as an env var into the container. That's to say, deployment line looks like this:

```bash
run: gcloud run deploy ${{ env.SERVICE }} --image="${{ env.GAR_LOCATION }}-docker.pkg.dev/${{ env.PROJECT_NAME }}/${{ env.REPOSITORY }}/${{ env.SERVICE }}:${{ github.sha }}" --region="${{ env.REGION }}" --env-vars-file=.env
```

Hope this helps!

**Reply by eliaspasche** [reply]
This results in an redirect loop if i run the container locally with http://localhost:3000/api/auth as `AUTH_URL`. Any ideas?

**Reply by kasp416h** [reply]
Also getting infinite redirect loop if i put anything in AUTH_URL or NEXTAUTH_URL in auth@beta version 30

**Comment by renanzan** [comment]
I'm facing the same issue. Currently, I'm using version "^5.0.0-beta.15" of next-auth with the Keycloak provider. Even after setting up all the suggested environment variables, I'm unable to proceed with authentication. Keycloak keeps reporting an invalid redirect_uri error, pointing to the Docker's internal host.

**Reply by teamsmiley** [reply]
hey , i had same issue, 
but i delete this 
```
# ENV HOSTNAME "0.0.0.0"
```
then it works please check if you have this.

**Comment by D3nn7** [comment]
We've updated today to Nextjs [14.2.2](https://github.com/vercel/next.js/releases/tag/v14.2.2) and removed the env `HOSTNAME` in Dockerfile. Now everything works as expected again.

**Reply by highree** [reply]
hey @D3nn7, 

can you share some insights ? I managed to run nextauth locally, as dev and build. As soon as i deploy on a dedicated server wiith https, it seems nothing happens. i get logged in i see the session for a second, and immediatly logged out. 

some insights how u get it running would be great.

**Reply by D3nn7** [reply]
@highree I have created a sample project that you can take a look at if you like :) Our production application and the sample project are working again. 

https://github.com/D3nn7/nextauth-discussion-8449

**Reply by highree** [reply]
thanks, i have the exact same config besides that i use siwe, and i found the solution there. seems the signOut declaration in the siwe config of web3modal caused an auto logout when session is there, but no network was connected. 

thanks for providing your sample project
cheers (oder prost in deinem fall)

**Comment by Qbason** [comment]
I got the same issue. On the website is written that AUTH_URL is not needed but without this I am redirecting to name of kubernetes pod instead to proper url website. I used dockerfile directly from next.js repo. I don't if this is important but I am also using AUTH_TRUST_HOST because of traefik.

Summary:
- Set AUTH_URL environment
- Use docker from https://github.com/vercel/next.js/blob/canary/examples%2Fwith-docker%2FDockerfile
- Set AUTH_TRUST_HOST environment if you behind reverse proxy

#Extra info
In new version of dockerfile the end command is different. I am using:

CMD ["node", "server.js"]

**Reply by johanwulf** [reply]
Yup, this is what did it for me as well. I've deployed my Next.js application to Google Cloud Run and spent some time scratching my head with this. `AUTH_URL=mydomain.com` and `AUTH_TUST_HOST=true` solved my problems.

**Reply by rylincoln** [reply]
This works for me in caprover deployment.

**Comment by benmarte** [comment]
I'm running into this issue as well I managed to get a bit further by adding the authorization option to the Keycloak options and now it tries to login but I still end up with the `ECONNREFUSED` error.

I made a repo with instructions on how to reproduce the issue in docker and how everything just works when you run nextjs locally and keycloak in docker.

Any help is appreciated, thanks.

https://github.com/benmarte/authjs-docker

**Reply by benmarte** [reply]
If anyone is using docker-compose and is still running into this issue I got it resolved thanks to the help of another keycloak user in their forums: https://keycloak.discourse.group/t/issues-running-keycloak-in-docker-and-authenticating-with-authjs/26443/14

**Comment by balazsorban44** [comment, marked as answer]
Hi everyone, we have an official section for deploying Docker here: https://authjs.dev/getting-started/deployment#docker

It includes the recommended Docker config, and can be seen deployed (running on Render.com, but it should work the same everywhere) here: https://nextjs-docker-example.authjs.dev

You should not need to set `AUTH_URL` `NEXTAUTH_URL`, `NEXTAUTH_URL_INTERNAL` and the like anymore.

**Reply by gnowland** [reply]
Unfortunately it's not the case -- running next-auth 5.0.0-beta.19 on next 14.2.4 any auth actions redirect to http://0.0.0.0 when running in Docker unless `AUTH_URL` is set or I use the `src/app/api/[...nextauth]/route.ts` `reqWithTrustedOrigin()` workaround as shown in this (closed) issue comment: https://github.com/nextauthjs/next-auth/issues/10928#issuecomment-2144241314. 

I hope this workaround can be added to core in the next patch so we don't have to add that to each project 🙏

**Reply by VladBo** [reply]
In my case the problem was caused by `NEXTAUTH_URL` and `NEXTAUTH_URL_INTERNAL` which were not removed from GCP Cloud Run instance after upgrade from NextAuth v4 to v5. They caused such a redirect loop. After 5 days of debugging I've finally fixed it 🫠
![2024-07-31 12 43 01](https://github.com/user-attachments/assets/131431cc-ffec-4fae-b002-60582ffd5573)

**Reply by ShivamAmin** [reply]
Just tried logging in with github on https://nextjs-docker-example.authjs.dev/ and I get a similar Configuration error (in my case its a redirect_uri_mismatch) when trying in my own app. I was able to successfully login to https://next-auth-example.vercel.app/. Setting the `AUTH_URL` env variable allows me to login without issue on my app.

**Reply by carrithers** [reply]
The advice here and in the docs is wrong. 

I am deploying my Nextjs server with ECS Fargate and next-auth@5.0.0-beta.25. My `host` and `x-forwarded-host` headers correctly point to my public domain, but setting `trustHost: true` does not work. The Dockerfile in the example repo is setting `HOSTNAME=0.0.0.0`, but all that does is change my callback domain to `0.0.0.0`. 

Eventually I simply set `AUTH_URL` and now it is working fine. Ironically, the example repo for the Docker deployment is also setting `AUTH_URL` in the docker-compose file, which probably explains why it works. 

https://github.com/nextauthjs/next-auth-example/blob/f2754cf5f37f100e126284702fbd1e6cc75e5afa/docker-compose.yml#L11

Either there is a bug or the docs need to be updated.

**Reply by skulden13** [reply]
Thanks! It works for me. I've also added AUTH_SECRET env variable

**Reply by dqnk** [reply]
Also had to set `AUTH_URL=https://subdomain.domain.tld` when self-hosting. Else it does not work.

**Reply by MaKTaiL** [reply]
> The advice here and in the docs is wrong.
> 
> I am deploying my Nextjs server with ECS Fargate and next-auth@5.0.0-beta.25. My `host` and `x-forwarded-host` headers correctly point to my public domain, but setting `trustHost: true` does not work. The Dockerfile in the example repo is setting `HOSTNAME=0.0.0.0`, but all that does is change my callback domain to `0.0.0.0`.
> 
> Eventually I simply set `AUTH_URL` and now it is working fine. Ironically, the example repo for the Docker deployment is also setting `AUTH_URL` in the docker-compose file, which probably explains why it works.
> 
> https://github.com/nextauthjs/next-auth-example/blob/f2754cf5f37f100e126284702fbd1e6cc75e5afa/docker-compose.yml#L11
> 
> Either there is a bug or the docs need to be updated.

Yeap, I have the same experience. To make it run with Docker we need to set AUTH_URL or else it won't work at all.

**Comment by Abhishek21k** [comment]
so while self hosting we don't need to specify the AUTH_URL?

**Reply by nmagnusson** [reply]
I had to use `AUTH_URL` for my self-hosted application. I use Docker and nginx.
In nginx I also had to add header Content-Security-Policy at `form-action` that the value was set to `'self' https://github.com`
